app.py

```diff
@@ -207,6 +207,12 @@ def initialize_session():
     login_user_from_token_cookie()
 
 
+@app.before_request
+def initialize_achievements():
+    if current_user()['username'] and 'achieved' not in session:
+        ACHIEVEMENTS.initialize_user_data(current_user()['username'])
```

Comment:
I'm not opposed to it, but doing it here depends on ordering. If in the same request a user logs in *and* is supposed to get an achievement (or not), using an `@app.before_request` requires that the code which sets `current_user()['username']` is executed in the same request before this code runs.

Did you verify that is the case? 

Otherwise, a simpler way to go would be to keep the first load completely a concern of the `ACHIEVEMENTS` class. It can easily see whether or not its data has already been initialized, and do the load whenever it's called.

Every method would start with a call to:

```py
self.initialize_session_if_necessary()
```

Reply:
I agree that it would be better to make it a concern of the `ACHIEVEMENTS` class. For now it all seems to work fine with `@app.before_request` as there is no achievements that can be achieved on the same request as login. As the highest priority is to get it correctly working before tomorrow I would like to leave it this way for now. I will create an issue to optimize the class overall, such as the storing of `commands` and `achievements` that can be optimized (as you suggested) through using maps and/or numbers for reference.